keyshelf-action v1.0.0: write-identity skips sops-only configs, sops decryption fails with ENOENT

## Summary

`keyshelf-action@v1.0.0`'s `write-identity.mjs` only writes the `identity` input to disk when the config declares a `provider:age` record. Configs that exclusively use `sops()` — even though `SopsProviderOptions` has its own `identityFile` field that `SopsProvider.readIdentity` reads at decrypt time — never get the identity file written. In CI this surfaces as every sops-encrypted key failing validation with `ENOENT: no such file or directory, open '<sopsProvider.identityFile>'`.

## Repro

`keyshelf.config.ts` (sops-only, no age providers anywhere):

```ts
import { config, defineConfig, secret, sops } from "keyshelf/config";

export default defineConfig({
  name: "sunsay-apps",
  envs: ["local", "staging", "production"],
  keys: {
    backoffice: {
      "cookie-signing-secret": secret({
        values: {
          local: sops({ identityFile: "~/.keyshelf/sunsay-apps/local.txt", secretsFile: "./.keyshelf/secrets/local.json" }),
          staging: sops({ identityFile: "~/.keyshelf/sunsay-apps/staging.txt", secretsFile: "./.keyshelf/secrets/staging.json" }),
          production: sops({ identityFile: "~/.keyshelf/sunsay-apps/production.txt", secretsFile: "./.keyshelf/secrets/production.json" }),
        },
      }),
    },
  },
});
```

Workflow:

```yaml
- uses: pantoninho/keyshelf/packages/action@b76b05f602e1b632c053b8af4d6b0ac3825c510b # keyshelf-action-v1.0.0
  with:
    env: staging
    identity: \${{ secrets.KEYSHELF_IDENTITY }}
    map: ./apps/x/.env.keyshelf
```

CI output:

```
##[warning]'identity' input was provided but config "sunsay-apps" declares no age providers. Ignoring.
##[error]Validation errors:
  - backoffice/cookie-signing-secret: ENOENT: no such file or directory, open '/home/runner/.keyshelf/sunsay-apps/staging.txt'
  ...
```

Real-world incident: sunsay-ltd/sunsay-apps#177 (v4 → v5 cutover); staging deploy run https://github.com/sunsay-ltd/sunsay-apps/actions/runs/25376913828/job/74414450273.

## Root cause

`packages/action/scripts/write-identity.mjs`:

```js
function visitBinding(binding) {
  if (binding === undefined) return;
  if (typeof binding !== "object" || binding === null) return;
  if (binding.__kind !== "provider:age") return;     // <-- skips provider:sops
  const file = binding.options?.identityFile;
  if (typeof file === "string" && file.length > 0) seen.add(file);
}
```

`SopsProviderOptions` has the same `identityFile` shape — it's just not collected. The asymmetry is invisible to users because the `identity` input description says \"Written to the path declared on the matching age provider\", but in practice sops uses age recipients under the hood and needs the same file on disk.

## Suggested fix

Have `collectAgeIdentityFiles` (or rename to `collectIdentityFiles`) also pick up `provider:sops` bindings whose `options.identityFile` is set. Since age and sops use the same identity material (an age private key) and the same path semantics, the write logic is identical.

Sketch:

```js
if (binding.__kind !== "provider:age" && binding.__kind !== "provider:sops") return;
const file = binding.options?.identityFile;
if (typeof file === "string" && file.length > 0) seen.add(file);
```

(Plus a test in `packages/action/test/scripts.test.ts` covering a sops-only config.)

## Workaround

Add an explicit `mkdir -p ~/.keyshelf/<name>/ && printf '%s\n' "\$KEYSHELF_IDENTITY" > ~/.keyshelf/<name>/<env>.txt && chmod 600` step before every `keyshelf-action` invocation. Removable once this is fixed and the action is re-pinned.

--- a/packages/action/action.yml
+++ b/packages/action/action.yml
@@ -13,7 +13,7 @@ inputs:
     description: Path(s) to .env.keyshelf mapping file(s); newline-separated for multiple
     required: true
   identity:
-    description: Raw identity material (e.g. age private key). Written to the path declared on the matching age provider with mode 0600. Pass via secrets.* — never hard-code.
+    description: Raw identity material (e.g. age private key). Written with mode 0600 to every identityFile path declared by age and sops providers in the config. Pass via secrets.* — never hard-code.
     required: false
     default: ""
   working-directory:

--- a/packages/action/scripts/identity-paths.mjs
+++ b/packages/action/scripts/identity-paths.mjs
@@ -1,24 +1,27 @@
 import { homedir } from "node:os";
 import { isAbsolute, join, resolve as pathResolve } from "node:path";
 
-export function ageIdentityFile(binding) {
-  if (binding?.__kind !== "provider:age") return undefined;
+const IDENTITY_PROVIDER_KINDS = new Set(["provider:age", "provider:sops"]);
+
+export function identityFile(binding) {
+  if (typeof binding !== "object" || binding === null) return undefined;
+  if (!IDENTITY_PROVIDER_KINDS.has(binding.__kind)) return undefined;
   const file = binding.options?.identityFile;
   return typeof file === "string" && file.length > 0 ? file : undefined;
 }
 
-export function collectAgeIdentityFiles(config) {
+export function collectIdentityFiles(config) {
   const seen = new Set();
   for (const record of config.keys) {
     if (record.kind !== "secret") continue;
-    addIfAge(seen, record.value);
-    for (const v of Object.values(record.values ?? {})) addIfAge(seen, v);
+    addIfIdentity(seen, record.value);
+    for (const v of Object.values(record.values ?? {})) addIfIdentity(seen, v);
   }
   return [...seen];
 }
 
-function addIfAge(seen, binding) {
-  const file = ageIdentityFile(binding);
+function addIfIdentity(seen, binding) {
+  const file = identityFile(binding);
   if (file !== undefined) seen.add(file);
 }
 

--- a/packages/action/scripts/write-identity.mjs
+++ b/packages/action/scripts/write-identity.mjs
@@ -3,7 +3,7 @@ import { writeFile, mkdir, chmod } from "node:fs/promises";
 import { dirname } from "node:path";
 import { loadConfig } from "keyshelf";
 import {
-  collectAgeIdentityFiles,
+  collectIdentityFiles,
   ensureTrailingNewline,
   resolveIdentityPath
 } from "./identity-paths.mjs";
@@ -17,11 +17,11 @@ if (!identity) {
 }
 
 const loaded = await loadConfig(cwd);
-const identityFiles = collectAgeIdentityFiles(loaded.config);
+const identityFiles = collectIdentityFiles(loaded.config);
 
 if (identityFiles.length === 0) {
   process.stdout.write(
-    `::warning::'identity' input was provided but config "${loaded.config.name}" declares no age providers. Ignoring.\n`
+    `::warning::'identity' input was provided but config "${loaded.config.name}" declares no providers that consume an identity file. Ignoring.\n`
   );
   process.exit(0);
 }

--- a/packages/action/test/identity-paths.test.ts
+++ b/packages/action/test/identity-paths.test.ts
@@ -2,53 +2,58 @@ import { describe, it, expect } from "vitest";
 import { homedir } from "node:os";
 import { join, isAbsolute } from "node:path";
 import {
-  ageIdentityFile,
-  collectAgeIdentityFiles,
+  identityFile,
+  collectIdentityFiles,
   ensureTrailingNewline,
   resolveIdentityPath
 } from "../scripts/identity-paths.mjs";
 
-describe("ageIdentityFile", () => {
+describe("identityFile", () => {
   it("returns the identityFile when binding is an age provider with a string path", () => {
-    expect(
-      ageIdentityFile({ __kind: "provider:age", options: { identityFile: "./key.txt" } })
-    ).toBe("./key.txt");
+    expect(identityFile({ __kind: "provider:age", options: { identityFile: "./key.txt" } })).toBe(
+      "./key.txt"
+    );
   });
 
-  it("returns undefined for non-age providers", () => {
+  it("returns the identityFile when binding is a sops provider with a string path", () => {
+    expect(identityFile({ __kind: "provider:sops", options: { identityFile: "./key.txt" } })).toBe(
+      "./key.txt"
+    );
+  });
+
+  it("returns undefined for providers that don't consume an identity file", () => {
     expect(
-      ageIdentityFile({ __kind: "provider:gcp", options: { identityFile: "./key.txt" } })
+      identityFile({ __kind: "provider:gcp", options: { identityFile: "./key.txt" } })
     ).toBeUndefined();
   });
 
   it("returns undefined for null/undefined/non-object bindings", () => {
-    expect(ageIdentityFile(null)).toBeUndefined();
-    expect(ageIdentityFile(undefined)).toBeUndefined();
-    expect(ageIdentityFile("string")).toBeUndefined();
-    expect(ageIdentityFile(42)).toBeUndefined();
+    expect(identityFile(null)).toBeUndefined();
+    expect(identityFile(undefined)).toBeUndefined();
+    expect(identityFile("string")).toBeUndefined();
+    expect(identityFile(42)).toBeUndefined();
   });
 
   it("returns undefined when identityFile is missing or empty", () => {
-    expect(ageIdentityFile({ __kind: "provider:age", options: {} })).toBeUndefined();
+    expect(identityFile({ __kind: "provider:age", options: {} })).toBeUndefined();
+    expect(identityFile({ __kind: "provider:age", options: { identityFile: "" } })).toBeUndefined();
     expect(
-      ageIdentityFile({ __kind: "provider:age", options: { identityFile: "" } })
+      identityFile({ __kind: "provider:age", options: { identityFile: 123 } })
     ).toBeUndefined();
-    expect(
-      ageIdentityFile({ __kind: "provider:age", options: { identityFile: 123 } })
-    ).toBeUndefined();
-    expect(ageIdentityFile({ __kind: "provider:age" })).toBeUndefined();
+    expect(identityFile({ __kind: "provider:age" })).toBeUndefined();
+    expect(identityFile({ __kind: "provider:sops", options: {} })).toBeUndefined();
   });
 });
 
-describe("collectAgeIdentityFiles", () => {
+describe("collectIdentityFiles", () => {
   it("returns empty array when there are no secret records", () => {
     const config = {
       keys: [{ kind: "config", path: "db/host" }]
     };
-    expect(collectAgeIdentityFiles(config)).toEqual([]);
+    expect(collectIdentityFiles(config)).toEqual([]);
   });
 
-  it("collects identityFile from a record's `value` binding", () => {
+  it("collects identityFile from a record's `value` binding (age)", () => {
     const config = {
       keys: [
         {
@@ -58,10 +63,26 @@ describe("collectAgeIdentityFiles", () => {
         }
       ]
     };
-    expect(collectAgeIdentityFiles(config)).toEqual(["./a.txt"]);
+    expect(collectIdentityFiles(config)).toEqual(["./a.txt"]);
   });
 
-  it("collects identityFile from per-env `values` bindings", () => {
+  it("collects identityFile from a record's `value` binding (sops)", () => {
+    const config = {
+      keys: [
+        {
+          kind: "secret",
+          path: "db/password",
+          value: {
+            __kind: "provider:sops",
+            options: { identityFile: "./s.txt", secretsFile: "./s.json" }
+          }
+        }
+      ]
+    };
+    expect(collectIdentityFiles(config)).toEqual(["./s.txt"]);
+  });
+
+  it("collects identityFile from per-env `values` bindings across age and sops", () => {
     const config = {
       keys: [
         {
@@ -69,26 +90,33 @@ describe("collectAgeIdentityFiles", () => {
           path: "db/password",
           values: {
             dev: { __kind: "provider:age", options: { identityFile: "./dev.txt" } },
-            prod: { __kind: "provider:age", options: { identityFile: "./prod.txt" } }
+            prod: {
+              __kind: "provider:sops",
+              options: { identityFile: "./prod.txt", secretsFile: "./prod.json" }
+            }
           }
         }
       ]
     };
-    expect(collectAgeIdentityFiles(config).sort()).toEqual(["./dev.txt", "./prod.txt"]);
+    expect(collectIdentityFiles(config).sort()).toEqual(["./dev.txt", "./prod.txt"]);
   });
 
-  it("deduplicates repeated paths across records and envs", () => {
-    const same = { __kind: "provider:age", options: { identityFile: "./shared.txt" } };
+  it("deduplicates repeated paths across records, envs, and provider kinds", () => {
+    const age = { __kind: "provider:age", options: { identityFile: "./shared.txt" } };
+    const sops = {
+      __kind: "provider:sops",
+      options: { identityFile: "./shared.txt", secretsFile: "./s.json" }
+    };
     const config = {
       keys: [
-        { kind: "secret", path: "a", value: same },
-        { kind: "secret", path: "b", value: same, values: { dev: same } }
+        { kind: "secret", path: "a", value: age },
+        { kind: "secret", path: "b", value: sops, values: { dev: age } }
       ]
     };
-    expect(collectAgeIdentityFiles(config)).toEqual(["./shared.txt"]);
+    expect(collectIdentityFiles(config)).toEqual(["./shared.txt"]);
   });
 
-  it("ignores non-age provider bindings and config keys", () => {
+  it("ignores provider bindings that don't consume an identity file", () => {
     const config = {
       keys: [
         { kind: "config", path: "db/host", value: "localhost" },
@@ -99,7 +127,7 @@ describe("collectAgeIdentityFiles", () => {
         }
       ]
     };
-    expect(collectAgeIdentityFiles(config)).toEqual([]);
+    expect(collectIdentityFiles(config)).toEqual([]);
   });
 });
 

--- a/packages/action/test/scripts.test.ts
+++ b/packages/action/test/scripts.test.ts
@@ -81,7 +81,7 @@ describe("write-identity.mjs", () => {
     expect((await stat(target)).mode & 0o777).toBe(0o600);
   });
 
-  it("warns and skips when the config declares no age providers", async () => {
+  it("warns and skips when the config declares no providers that consume an identity file", async () => {
     const root = await mkdtemp(join(tmpdir(), "keyshelf-action-noage-"));
     await writeFile(
       join(root, "keyshelf.config.ts"),
@@ -104,7 +104,47 @@ describe("write-identity.mjs", () => {
 
     expect(res.status).toBe(0);
     expect(res.stdout).toContain("::warning::");
-    expect(res.stdout).toContain("declares no age providers");
+    expect(res.stdout).toContain("no providers that consume an identity file");
+  });
+
+  it("writes identity for sops-only configs", async () => {
+    const root = await mkdtemp(join(tmpdir(), "keyshelf-action-sops-"));
+    const identity = await generateIdentity();
+
+    await writeFile(
+      join(root, "keyshelf.config.ts"),
+      [
+        'import { defineConfig, secret, sops } from "keyshelf/config";',
+        "",
+        "export default defineConfig({",
+        '  name: "sops-only",',
+        '  envs: ["test"],',
+        "  keys: {",
+        "    db: {",
+        "      password: secret({",
+        "        values: {",
+        '          test: sops({ identityFile: "./keys/sops.txt", secretsFile: "./.keyshelf/secrets/test.json" })',
+        "        }",
+        "      })",
+        "    }",
+        "  }",
+        "});"
+      ].join("\n")
+    );
+
+    const res = spawnSync(process.execPath, [WRITE_IDENTITY], {
+      cwd: root,
+      env: { ...process.env, KEYSHELF_IDENTITY: identity, KEYSHELF_CWD: root },
+      encoding: "utf-8"
+    });
+
+    expect(res.status, res.stderr).toBe(0);
+    expect(res.stdout).toContain("Wrote identity to");
+    expect(res.stdout).toContain("keys/sops.txt");
+
+    const target = join(root, "keys/sops.txt");
+    expect((await readFile(target, "utf-8")).trim()).toBe(identity.trim());
+    expect((await stat(target)).mode & 0o777).toBe(0o600);
   });
 
   it("is a no-op when no identity is provided", async () => {

